---
title: "Sperm short RNA-Seq data analysis"
output:
  html_document:
    theme: cerulean
    df_print: paged
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Library
```{r sperm-1, message=FALSE, warning=FALSE}
suppressPackageStartupMessages({
  library(TreeSummarizedExperiment)
  library(treeclimbR)
  library(ape)
  library(ggtree)
  library(ggplot2)
  library(cowplot)
  library(edgeR)
  library(ComplexHeatmap)
  library(dplyr)
  library(TreeHeatmap)
  library(ggnewscale)
  library(scales)
  library(castor)
})
```

# TSE
```{r}
tse <- readRDS("../03_tse/tse_mir.rds")
colnames(colData(tse)) <- tolower(colnames(colData(tse)))

dea <- runDA(TSE = tse)

out <- nodeResult(object = dea, n = Inf)

d <- cpm(
  calcNormFactors(
    DGEList(counts = assay(tse), samples = colData(tse))
  ),
  log = TRUE
)

assays(tse)[2] <- list(cpm = d)


cand <- getCand(
  tree = rowTree(tse),
  score_data = out, node_column = "node",
  p_column = "PValue", sign_column = "logFC",
  message = FALSE
)

candB <- evalCand(
  tree = rowTree(tse),
  levels = cand$candidate_list,
  score_data = out, node_column = "node",
  p_column = "PValue", sign_column = "logFC",
  method = "BH", limit_rej = 0.05,
  use_pseudo_leaf = FALSE,
  message = FALSE
)

result <- topNodes(object = candB, n = Inf, p_value = 1)


# all nodes
all_node <- showNode(tree = rowTree(tse), only.leaf = FALSE)

# Tree
tree_mir <- get_subtree_at_node(tree = rowTree(tse), node = "miRNA")$subtree
fig_tree <- ggtree(tree_mir,
  ladderize = FALSE,
  layout = "fan",
  open.angle = 10,
  size = 0.5
)

loc_0.05 <- result$node
names(loc_0.05) <- ifelse(result$logFC > 0, "Up-regulated", "Down-regulated")


br_0.05 <- findOS(
  tree = rowTree(tse), node = loc_0.05,
  only.leaf = FALSE, self.include = TRUE
)

df_branch <- data.frame(
  node = unlist(br_0.05),
  Branch = rep(
    names(loc_0.05),
    unlist(lapply(br_0.05, length))
  )
) %>%
  distinct() %>%
  add_row(
    node = setdiff(all_node, unlist(br_0.05)),
    Branch = "same"
  )

# branch colors
branch_col <- c(
  "same" = "grey70", "Up-regulated" = "navy",
  "Down-regulated" = "orange"
)

# figure: tree
fig_tree <- fig_tree %<+%
  df_branch +
  aes(color = Branch) +
  scale_color_manual(
    values = branch_col,
    guide = guide_legend(
      override.aes = list(size = 3),
      order = 1
    )
  ) +
  geom_point2(aes(subset = (node %in% loc_0.05)),
    color = "red", shape = 19, size = 1.5
  )

fig_tree
```


#Scale branches
```{r}
# branches without signal
leaf <- showNode(tree = rowTree(tse), only.leaf = TRUE)
leaf_yes <- findOS(
  tree = rowTree(tse), node = loc_0.05,
  only.leaf = TRUE, self.include = TRUE
)
leaf_no <- setdiff(leaf, unlist(leaf_yes))
node_no <- signalNode(tree = rowTree(tse), node = leaf_no)
node_no <- node_no[!isLeaf(tree = rowTree(tse), node = node_no)]


circo_full <- fig_tree
for (i in node_no) {
  circo_full <- scaleClade(circo_full, node = i, scale = 0.2)
}
circo_full
```

# Label hierarchical levels
```{r}
ux <- setdiff(sort(unique(circo_full$data$x)), 0)
# names(ux) <- c("genomic_cluster", "primary_tx", "miRNA", "sequence" )
df_layer <- data.frame(x = ux) %>%
  arrange(x) %>%
  mutate(
    y = max(circo_full$data$y, na.rm = TRUE),
    layer = factor(ux)
  )

# annotate the four levels on the tree
layer_shape <- 49:55
names(layer_shape) <- ux
circo_full <- circo_full +
  geom_point(
    data = df_layer,
    aes(x, y = y + 210, shape = layer),
    size = 6,
    inherit.aes = FALSE, show.legend = FALSE
  ) +
  scale_shape_manual(
    values = layer_shape,
    guide = guide_legend(
      override.aes = list(size = 4),
      order = 1
    )
  ) +
  labs(shape = "Hierarchy")
circo_full
```

# Heatmap
```{r}
yes_L <- findOS(
  tree = rowTree(tse), node = loc_0.05,
  only.leaf = TRUE, self.include = TRUE
)
yes_L <- rowLinks(tse)$nodeNum %in% unlist(yes_L)

# scale counts:
#   1. log transformed in cpm counts
#   2. log-count of each miRNA is further scaled among samples

# log-transformed by columns
count_L <- assays(tse)[[1]][yes_L, ]
countL <- cpm(count_L)
scale_L <- assays(tse)[[2]][yes_L, ]

# standardized
scale_L <- t(apply(scale_L, 1, FUN = function(y) {
  ly <- length(unique(y))
  if (ly == 1) {
    sy <- rep(0, length(y))
  } else {
    sy <- scale(y)
  }
  return(sy)
}))
rownames(scale_L) <- rownames(count_L)
colnames(scale_L) <- colnames(count_L)
dim(scale_L)
```


```{r}
group <- colData(tse)$group
names(group) <- colnames(tse)

fig2 <- TreeHeatmap(
  tree = rowTree(tse),
  rel_width = 0.2,
  tree_fig = circo_full,
  hm_data = scale_L,
  legend_title_hm = "Expression(scale)",
  tree_hm_gap = 0.3,
  column_split = group,
  column_split_gap = 0.05,
  column_split_label = c(
    "Group1" = "Group1",
    "Group2" = "Group2"
  ),
  split_label_angle = 80,
  split_label_size = 4,
  split_label_offset_y = 260
) +
  scale_fill_gradient(
    low = "navy", high = "yellow",
    limits = c(-1, 1),
    oob = squish,
    guide = guide_colorbar(order = 3)
  ) +
  new_scale_fill()
```

```{r}
leaf_a <- findOS(
  tree = rowTree(tse), node = loc_0.05,
  only.leaf = TRUE, self.include = TRUE
)

names(leaf_a) <- transNode(
  tree = rowTree(tse), node = loc_0.05,
  use.alias = TRUE
)

df_a <- lapply(loc_0.05, FUN = function(x) {
  xx <- rowData(tse)[rowLinks(tse)$nodeNum %in% x, ]
})


mir_a <- lapply(loc_0.05, FUN = function(x) {
  ix <- (rowLinks(tse)$nodeNum %in% x) & rowLinks(tse)$isLeaf
  xx <- rowData(tse)[ix, "Annotation"]
  unique(as.character(xx))
})
head(mir_a)
```












# Simulated data
```{r}
tse <- readRDS("../03_tse/tse.rds")
colnames(colData(tse)) <- tolower(colnames(colData(tse)))

d <- cpm(
  calcNormFactors(
    DGEList(counts = assay(tse), samples = colData(tse))
  ),
  log = TRUE
)

assays(tse)[[2]] <- d



sim <- rownames(readRDS("../../data/simulated/readsToSpike.rds"))
tse_sim <- tse[rownames(tse) %in% sim, ]
tree_mir <- get_subtree_at_node(tree = rowTree(tse), node = "miRNA")$subtree

tree_sim <- get_subtree_with_tips(tree = tree_mir, only_tips = sim)$subtree

p_mir <- ggtree(tree_sim, color = "firebrick", size = 1, branch.length = "none", type = "fan") +
  geom_nodelab(size = 3, color = "purple", angle = 0, vjust = -1.5, hjust = 1.5) +
  geom_tiplab(size = 10, color = "black", as_ylab = TRUE)

fig_tree <- ggtree(tree_sim,
  ladderize = FALSE,
  layout = "fan",
  open.angle = 10,
  size = 0.5
)


m <- assays(tse)[[2]][tree_sim$tip.label, ]
m <- m-rowMeans(m[,1:4])

TreeHeatmap(
  tree = tree_sim,
  tree_fig = p_mir,
  tree_hm_gap = 0.5,
  rel_width = 0.5,
  hm_data = m,
  legend_title_hm = expression(Log[2]~FC~"(expression)"),
) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red",
                        limits = c(-2, 2),
                        oob = squish,
                        guide = guide_colorbar(order = 3))



# result <- topNodes(object = dea, n = Inf, p_value = 0.05)
# loc_0.05 <- result$node


p <- ggtree(tree_sim, color = "firebrick", size = 1, branch.length = "none") +
  geom_nodelab(size = 3, color = "purple", angle = 0, vjust = -1.5, hjust = 1.5) +
  geom_tiplab(size = 10, color = "black", as_ylab = TRUE)

# viewClade(p, MRCA(p, "miR-99b"))

TreeHeatmap(
  tree = tree_sim,
  tree_fig = p,
  hm_data = assays(tse)[[2]][tree_sim$tip.label, ],
)
```






<!-- # Test -->
<!-- ```{r} -->
<!-- all_node <- showNode(tree = rowTree(tse), only.leaf = FALSE) -->

<!-- # the result table -->
<!-- # result <- topNodes(object = candB, n = Inf, p_value = 0.05) -->
<!-- result <- out[abs(out$logFC) >= 1 & out$PValue <= 0.05, ] -->
<!-- loc_0.05 <- result$node -->

<!-- # the logFC of reported nodes -->
<!-- loc_lfc <- result$logFC -->

<!-- # the sign of reported nodes -->
<!-- loc_sign <- ifelse(sign(loc_lfc) > 0, "Up-regulated", "Down-regulated") -->
<!-- names(loc_0.05) <- loc_sign -->

<!-- br_0.05 <- findOS( -->
<!--   tree = rowTree(tse), node = loc_0.05, -->
<!--   only.leaf = FALSE, self.include = TRUE -->
<!-- ) -->
<!-- df_branch <- data.frame( -->
<!--   node = unlist(br_0.05), -->
<!--   Branch = rep( -->
<!--     names(loc_0.05), -->
<!--     unlist(lapply(br_0.05, length)) -->
<!--   ) -->
<!-- ) %>% -->
<!--   distinct() %>% -->
<!--   add_row( -->
<!--     node = setdiff(all_node, unlist(br_0.05)), -->
<!--     Branch = "same" -->
<!--   ) -->

<!-- # branch colors -->
<!-- branch_col <- c( -->
<!--   "same" = "grey70", "Up-regulated" = "navy", -->
<!--   "Down-regulated" = "orange" -->
<!-- ) -->

<!-- # figure: tree -->
<!-- fig_tree <- fig_tree %<+% -->
<!--   df_branch + -->
<!--   aes(color = Branch) + -->
<!--   scale_color_manual( -->
<!--     values = branch_col, -->
<!--     guide = guide_legend( -->
<!--       override.aes = list(size = 3), -->
<!--       order = 1 -->
<!--     ) -->
<!--   ) + -->
<!--   geom_point2(aes(subset = (node %in% loc_0.05)), -->
<!--     color = "red", shape = 19, size = 1.5 -->
<!--   ) -->

<!-- fig_tree -->
<!-- ``` -->















# Simulated reads
```{r}
sim <- readRDS("../../data/simulated/readsToSpike.rds")
res <- readRDS("results.rds")
res_sim <- res[rownames(res) %in% rownames(sim), ]
res_sim$logfc_real <- "2"
res_sim["AAGTTTGTGATGAGACACTAC", "logfc_real"] <- 0
res_sim["CTGTAAGTTTGTGATGAGAC", "logfc_real"] <- 0

cnts <- cpm(calcNormFactors(DGEList(assay(tse)[rownames(res_sim), ])), log = T)
cnts <- cnts - rowMeans(cnts[, 1:4])


set.seed(123)
hm <- Heatmap(
  matrix = cnts,
  border = T,
  cluster_rows = F,
  cluster_columns = F,
  top_annotation = HeatmapAnnotation(Groups = colData(tse)[, 2], annotation_name_gp = gpar(fontsize = 0)),
  left_annotation = rowAnnotation(
    Biotype = as.character(res_sim[, c("transcript_type")]),
    annotation_name_gp = gpar(fontsize = 0)
  ),
  row_split = as.character(res_sim[, c("transcript_type")]),
  column_split = colData(tse)[, 2],
  width = unit(6, "cm"),
  height = unit(6, "cm"),
  row_title_gp = gpar(fontsize = 0),
  row_names_gp = gpar(fontsize = 10),
  column_names_gp = gpar(fontsize = 10),
  heatmap_legend_param = list(
    at = c(-4, 0, 4),
    title = expression(Log[2] ~ "FC (CPM)"),
    legend_height = unit(1, "cm")
  )
) + Heatmap(
  matrix = res_sim[, "logFC"],
  cluster_rows = FALSE,
  width = unit(0.5, "cm"),
  name = "Log2FC",
  column_names_gp = gpar(fontsize = 0),
  border = TRUE,
  row_labels = rownames(res_sim),
  row_names_gp = gpar(fontsize = 10),
  heatmap_legend_param = list(
    at = c(-2, 0, 2),
    title = expression(Log[2] ~ "FC"),
    legend_height = unit(1, "cm")
  )
) +
  Heatmap(
    matrix = res_sim[, "logfc_real"],
    cluster_rows = FALSE,
    width = unit(0.5, "cm"),
    name = "LogFC",
    column_names_gp = gpar(fontsize = 0),
    border = TRUE,
    row_labels = rownames(res_sim),
    row_names_gp = gpar(fontsize = 10),
    heatmap_legend_param = list(
      at = c(0, 2),
      title = expression("Simulated" ~ Log ~ "FC"),
      legend_height = unit(1, "cm")
    )
  )



draw(hm,
  heatmap_legend_side = "bottom",
  annotation_legend_side = "right",
  merge_legends = TRUE,
  align_annotation_legend = ""
)
```
